doc/source/development/rolling_operations_with_numba.rst

```diff
@@ -0,0 +1,234 @@
+Rolling Operations with Numba
+=============================
+
+Summary
+-------
+
+This proof of concept (POC) demonstrates using `Numba <http://numba.pydata.org/>`_ instead of ``Cython``
+to compute ``rolling.mean`` and ``rolling.apply`` without introducing any user facing API changes.
+The benefits of using ``Numba`` include:
+
+#. Performance parity or improvement over ``Cython``
+#. Eliminate shipping C-extensions
+#. Ease of debugging and maintenance as pure Python code
+
+Additionally this POC includes an new API that allows users to create custom window boundary calculations
+when using ``Numba``.
+
+Immediate Proposal
+~~~~~~~~~~~~~~~~~~
+
+In line with the `pandas roadmap <https://pandas.pydata.org/pandas-docs/stable/development/roadmap.html#numba-accelerated-operations>`_,
+we would like to introduce ``Numba`` as a required dependency for pandas in the 1.0 release
+and have the ``rolling.mean`` and ``rolling.apply`` operations use ``Numba`` instead of ``Cython`` in the release.
+
+Implementation
+--------------
+
+``rolling.mean`` and ``rolling.apply`` utilizes ``njit`` functions to separately calculate:
+
+* Window boundaries based on the window size and index
+* The actual aggregation function (mean and apply in this POC)
+
+The functional implementation mimics the current ``Cython`` implementation; however, now that
+the calculation of the window boundaries is separate from the aggregation function, we are able to
+expose a new API to users that allows them to specify how to calculate the rolling window boundaries.
+
+New Custom Window Indexer API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Currently, window bounds are calculated automatically based on the ``DataFrame`` or ``Series`` index
+when a user passes an integer or offset in the rolling API (e.g. ``df.rolling(2)``). This POC also introduces
+a ``BaseIndexer`` class in ``pandas/core/window/indexers.py`` for users to subclass
+to create a custom routine to calculate window boundaries. Users will need to specify a
+``get_window_bounds`` function to calculate window boundaries.
+
+Below is an example of creating an ``ExpandingIndexer`` that computes an expanding window similar to
+the current ``expanding`` API:
+
+.. code-block:: python
+
+   from typing import Optional
+
+   from pandas import Series
+   from pandas.core.window.indexers import BaseIndexer
```

Comment:
create a separate issue, we may want to expose the name in pandas.api.indexers (BaseIndexer) for user consumption.